Docs:  Loki data sources updates (#72041)
grafana/grafana: docs/sources/datasources/loki/query-editor/index.md

| {{< figure src="/static/img/docs/v95/loki_operation_description.png" class="docs-image--no-shadow" max-width="30px" >}} | Opens the operation's description tooltip.                        |
| {{< figure src="/static/img/docs/v95/loki_operation_remove.png" class="docs-image--no-shadow" max-width="30px" >}}      | Removes the operation.                                            |

The query editor groups operations into the following sections:

- Aggregations - see [Built-in aggregation operators](/docs/loki/latest/logql/metric_queries/#built-in-aggregation-operators)
- Range functions - see [Range Vector aggregation](/docs/loki/latest/logql/metric_queries/#range-vector-aggregation)
- Formats - see [Log queries](/docs/loki/latest/logql/log_queries/#log-queries)
- Binary operations - see [Binary operators](/docs/loki/latest/logql/#binary-operators)
- Label filters - see [Label filter expression](/docs/loki/latest/logql/log_queries/#label-filter-expression)
- Line filters - see [Line filter expression](/docs/loki/latest/logql/log_queries/#label-filter-expression)

> Some operations make sense only when used in a specific order.b If adding an operation would result in nonsensical query, the query editor adds the operation to the correct place.

Reviewer: ```suggestion
Some operations make sense only when used in a specific order. If adding an operation would result in nonsensical query, the query editor adds the operation to the correct place.
```

Reviewer: Does the query builder do this @ivanahuckova  ?

Reviewer 2: Yes - you can test it out by creating query such as `{a="b"} | logfmt` in builder and adding line filter operation. It will be placed before the parser.

To re-order operations manually, drag the operation box by its name and drop it into the desired place. For additional information see [Order of operations](/docs/loki/latest/logql/#order-of-operations).

### Hints

In same cases the query editor can detect which operations would be most appropriate for a selected log stream. In such cases it will show a hint next to the `+ Operations` button. Click on the hint to add the operations to your query.

## Code mode

> In **Code mode**, you can write complex queries using a text editor with autocompletion features and syntax highlighting.

Reviewer: ```suggestion
In **Code mode**, you can write complex queries using a text editor with autocompletion feature, syntax highlighting, and query validation.
```

It also contains a [label browser]({{< relref "#label-browser" >}}) to further help you write queries.

For more information about Loki's query language, refer to the [Loki documentation](/docs/loki/latest/logql/).

### Use autocompletion

Code mode's autocompletion feature works automatically while typing.

The query editor can autocomplete static functions, aggregations, and keywords, and also dynamic items like labels.
The autocompletion dropdown includes documentation for the suggested items where available.

## Options

The following options are the same for both **Builder** and **Code** mode:

- **Legend** - Controls the time series name, using a name or pattern. For example, `{{hostname}}` is replaced with the label value for the label `hostname`.

- **Type** - Selects the query type to run. The `instant` type queries against a single point in time. We use the "To" time from the time range. The `range` type queries over the selected range of time.

- **Line limit** -Defines the upper limit for the number of log lines returned by a query. The default is `1000`

> - **Resolution** Sets the step parameter of Loki metrics range queries. With a resolution of `1/1`, each pixel corresponds to one data point. `1/2` retrieves one data point for every other pixel, `1/10` retrieves one data point per 10 pixels, and so on. Lower resolutions perform better.

Reviewer 2: `Resolution` has been deprecated and `Step` has been added.
https://grafana.com/docs/grafana/next/datasources/loki/query-editor/#configure-query-settings

```suggestion
- **Step** Sets the step parameter of Loki metrics queries. The default value equals to the value of `$__interval` variable, which is calculated using the time range and the width of the graph (the number of pixels).

- **Resolution** Deprecated. Sets the step parameter of Loki metrics range queries. With a resolution of `1/1`, each pixel corresponds to one data point. `1/2` retrieves one data point for every other pixel, `1/10` retrieves one data point per 10 pixels, and so on. Lower resolutions perform better.
```
